MetaMask/metamask-extension
# feat: Add RPC Method Middleware params transform and tracking

## **Description**

Currently we track a sample of most RPC method calls, but we do not include params. This is for good reason of course since many param values are sensitive and may identify an otherwise anonymous MetaMask Metrics user. There are some param values that are not sensitive yet are not tracked as well. Our goal is to glean additional insight into the usage of our RPC methods by including non-sensitive params in the events that track them. 

This PR achieves that by including a map of method to transformation functions which accepts a request's params object (or array) and transforms it something that is acceptable for analytics. By default no params are included.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23236?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2167

## **Manual testing steps**

1. Opt into MetaMetrics in the wallet settings
1. Open your the background console for extension
1. Add a breakpoint for trackEvent inside `createRPCMethodTrackingMiddleware.js`
1. On any page with a provider injected, use `window.ethereum.request` to make some rpc method calls
1. Check background console for breakpoint trigger
  * Currently only `wallet_watchAsset` should include the `type` param. All other methods should not include any params.  
  * [the wallet_watchAsset in API playground](https://docs.metamask.io/wallet/reference/wallet_watchasset/) may be helpful to use

NOTE: I recommend breakpoint and debugger over the network tab because I don't believe the tracking requests actually fire on dev.

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
